src/main/java/it/unitn/ds1/Replica.java

```diff
@@ -230,203 +434,356 @@ private void onWriteMsg(WriteMsg msg) {
      */
     private void onWriteOkMsg(WriteOkMsg msg) {
         // If the epoch of the write is not the current epoch, ignore the message
-        if (msg.epoch != this.epoch)
+        if (msg.id.epoch != this.epoch)
+            // Resetting last contact here would be wrong since the received
+            // message comes from an already crashed coordinator
             return;
 
         // If the pair epoch-index is not in the map, ignore the message
-        var pair = new AbstractMap.SimpleEntry<>(msg.epoch, msg.writeIndex);
-        if (!this.writeRequests.containsKey(pair))
+        if (!this.writeRequests.containsKey(msg.id))
             return;
 
-        // If the last write applied is newer than the current write, ignore the message
-        // If we have already applied the write, ignore the message
-        if (this.lastUpdateApplied.epoch > msg.epoch || (
-                this.lastUpdateApplied.epoch == msg.epoch &&
-                        (this.lastUpdateApplied.writeIndex > msg.writeIndex ||
-                                this.lastUpdateApplied.writeIndex == msg.writeIndex)
-        )) return;
+        // Registers the arrival of the ok for a later check
+        this.writeOks.add(msg.id);
+
+        // Checks if this replicas has to inform the original client of the
+        // completed update [Must be done regardless of the subsequent check on
+        // request age to avoid wrong crash detection from the client]
+        if (this.updateRequests.contains(msg.updateRequestId)) {
+            this.updateRequests.remove(msg.updateRequestId);
```

Comment:
Why are we sending it only when we receive the OK from the coordinator? The client should only need to know that the replica received it, and that eventually it will be applied.
By waiting for the WriteOk, when an election occur it's easy for timeouts to happen.
Can we move it to onUpdateRequest?

Reply:
(I was referring to the tellDelayed to the client, happening on the following line)